hypershift-operator/controllers/nodepool/nodepool_controller.go

```diff
@@ -2621,21 +2645,37 @@ func (r *NodePoolReconciler) ensureMachineDeletion(nodePool *hyperv1.NodePool) e
 	return nil
 }
 
-// getMachinesForNodePool get all Machines listed with the nodePoolAnnotation
-// within the control plane Namespace for that NodePool.
-func (r *NodePoolReconciler) getMachinesForNodePool(nodePool *hyperv1.NodePool) ([]*capiv1.Machine, error) {
+// listSortedSortedMachinesForCluster get all Machines within the control plane Namespace for
+// the hosted cluster sorted by creation timestamp
+func (r *NodePoolReconciler) listSortedMachinesForCluster(hc *hyperv1.HostedCluster) ([]capiv1.Machine, error) {
+	machinesForCluster := capiv1.MachineList{}
+
+	if err := r.List(context.Background(), &machinesForCluster, &client.ListOptions{
+		Namespace: manifests.HostedControlPlaneNamespace(hc.Namespace, hc.Name),
+	}); err != nil {
+		return nil, fmt.Errorf("failed to list Machines: %w", err)
+	}
+
+	return sortedByCreationTimestamp(machinesForCluster.Items), nil
+}
```

Comment:
if we're going to be processing all machines for a cluster, then it seems like this belongs in a controller that has a cluster view.

is there a reason why this logic shouldn't go into the HostedCluster controller so it's reconciled only once when the hosted cluster is reconciled? That seems to map better to the logic than reconciling the same clusterwide object for every nodepool. the HC controller already has the kv infra client map too i believe. so it should be pretty straight forward.

Reply:
If hc controller Reconcile on nodepool changes we are fine, I will try that.